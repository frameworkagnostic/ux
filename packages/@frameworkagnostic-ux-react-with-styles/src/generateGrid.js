

export const GUTTERS = [0, 8, 16, 24, 32, 40];
export const GRID_SIZES = ['auto', true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

export function generateGrid(globalStylesArg, theme, breakpoint) {
  const styles = {};
  const globalStyles = globalStylesArg;

  GRID_SIZES.forEach((size) => {
    const key = `grid-${breakpoint}-${size}`;

    if (size === true) {
      // For the auto layouting
      styles[key] = {
        flexBasis: 0,
        flexGrow: 1,
        maxWidth: '100%',
      };
      return;
    }

    if (size === 'auto') {
      styles[key] = {
        flexBasis: 'auto',
        flexGrow: 0,
        maxWidth: 'none',
      };
      return;
    }

    // Only keep 6 significant numbers.
    const width = `${Math.round((size / 12) * 10e6) / 10e4}%`;

    // Close to the bootstrap implementation:
    // https://github.com/twbs/bootstrap/blob/8fccaa2439e97ec72a4b7dc42ccc1f649790adb0/scss/mixins/_grid.scss#L41
    styles[key] = {
      flexBasis: width,
      flexGrow: 0,
      maxWidth: width,
    };
  });

  // No need for a media query for the first size.
  if (breakpoint === 'xs') {
    Object.assign(globalStyles, styles);
  } else {
    globalStyles[theme.breakpoints.up(breakpoint)] = styles;
  }
}
