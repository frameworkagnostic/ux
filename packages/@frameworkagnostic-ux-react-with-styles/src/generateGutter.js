
import { GUTTERS } from './constants';

export function generateGutter(breakpoint) {
  const styles = {};

  GUTTERS.forEach((spacing, index) => {
    if (index === 0) {
      // Skip the default style.
      return;
    }

    styles[`spacing-${breakpoint}-${spacing}`] = {
      margin: -spacing / 2,
      width: `calc(100% + ${spacing}px)`,
      '& > $item': {
        padding: spacing / 2,
      },
    };
  });

  return styles;
}
