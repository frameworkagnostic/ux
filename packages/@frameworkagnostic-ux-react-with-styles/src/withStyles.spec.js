
import React from 'react';
import renderer from 'react-test-renderer';
import { withStyles } from './withStyles';
import { generateGutter } from './generateGutter';
import { generateGrid } from './generateGrid';
import { keys as breakpointKeys } from './createBreakpoints';

test('withStyles', () => {
  const Grid = ({ classes }) => <div>Hello</div>;
  const gutter = generateGutter('xs');

  const styles = theme => {
    expect(typeof theme).toBe('object');
    return {
      container: {
        boxSizing: 'border-box',
        display: 'flex',
        flexWrap: 'wrap',
        width: '100%',
      },
      item: {
        boxSizing: 'border-box',
        margin: '0',
      },
      ...gutter,
      ...breakpointKeys.reduce((accumulator, key) => {
        // Use side effect over immutability for better performance.
        generateGrid(accumulator, theme, key);
        return accumulator;
      }, {}),
    };
  };

  const StyledGrid = withStyles(styles, { name: 'MuiGrid' })(Grid);
  expect(typeof StyledGrid).toBe('function');
  expect(typeof StyledGrid.propTypes).toBe('object');
  const element = <StyledGrid />;
  const component = renderer.create(element);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})
