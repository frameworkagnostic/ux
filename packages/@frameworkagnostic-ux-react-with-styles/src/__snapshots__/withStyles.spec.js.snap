// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withStyles 1`] = `
<div>
  Hello
</div>
`;
